lightning/src/ln/outbound_payment.rs

```diff
@@ -902,15 +902,64 @@ impl OutboundPayments {
 			}
 		}
 
-		let mut route = router.find_route_with_id(
-			&node_signer.get_node_id(Recipient::Node).unwrap(), &route_params,
+		let payer = node_signer.get_node_id(Recipient::Node).unwrap();
+		let mut route;
+		match router.find_route_with_id(
+			&payer,&route_params,
 			Some(&first_hops.iter().collect::<Vec<_>>()), inflight_htlcs(),
 			payment_hash, payment_id,
-		).map_err(|_| {
-			log_error!(logger, "Failed to find route for payment with id {} and hash {}",
-				payment_id, payment_hash);
-			RetryableSendFailure::RouteNotFound
-		})?;
+		) {
+			Ok(res) => route = res,
+			Err(error) => {
+				if error.err == "Cannot generate a route to ourselves" {
+					let payment_secret = match recipient_onion.payment_secret {
+						Some(secret) => secret,
+						None => PaymentSecret([0; 32])
+					};
+					let payment_preimage = PaymentPreimage([0; 32]);
+					let payment_purpose = PaymentPurpose::InvoicePayment {
+						payment_preimage: Some(payment_preimage),
+						payment_secret,
+					};
+					let dummy_route = Route {
+						paths: vec![Path {
+							hops: vec![RouteHop {
+									pubkey: payer,
+									node_features: NodeFeatures::empty(),
+									short_channel_id: 0,
+									channel_features: ChannelFeatures::empty(),
+									fee_msat: 0,
+									cltv_expiry_delta: 0,
+									maybe_announced_channel: false,
+								}],
+							blinded_tail: None,
+						}],
+						route_params: Some(route_params.clone()),
+					};
+					let _ = self.add_new_pending_payment(payment_hash,
```

Comment:
It's a little unintuitive why we add a pending payment only to immediately mark it fulfilled. Is the reason solely for duplicate payment protection? Could we add a comment?

Reply:
Yes, my thinking was to protect against duplicates. Yes, will definitely add a comment.